runtime/src/main/java/org/corfudb/runtime/collections/StreamPollingTask.java

```diff
@@ -0,0 +1,108 @@
+
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.view.stream.IStreamView;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A runnable task for a subscription to poll data change from the
+ * transaction stream and put into the subscription's buffer. This
+ * task is executed by the thread pool continuously until an error
+ * occurs which would stop the subscription.
+ * <p>
+ * Created by WenbinZhu on 11/9/20.
+ */
+@Slf4j
+class StreamPollingTask implements Runnable {
+
+    // A period of time in ms to sleep before next cycle when poller gets no new data changes.
+    private static final Duration IDLE_WAIT_TIME_MS = Duration.ofMillis(50);
+
+    // Total amount of time to wait for putting the polled data changes into buffer if it is full.
+    private static final Duration QUEUE_FULL_BLOCK_TIME_MS = Duration.ofMillis(2_000);
+
+    // The streaming manager that is in charge of listener subscriptions.
+    private final StreamingManager streamingManager;
+
+    // The corfu transaction stream to poll data changes from.
+    private final IStreamView txnStream;
+
+    // The subscription context associated with this task.
+    private final StreamSubscription subscription;
+
+    // The Thread pool for executing stream polling tasks.
+    private final ExecutorService pollingExecutor;
+
+    // Last address of the data successfully processed by the buffer.
+    private long lastReadAddress;
+
+    StreamPollingTask(StreamingManager streamingManager, long lastAddress,
+                      StreamSubscription subscription, ExecutorService executor) {
+        this.streamingManager = streamingManager;
+        this.subscription = subscription;
+        this.pollingExecutor = executor;
+        this.lastReadAddress = lastAddress;
+        this.txnStream = subscription.getTxnStream();
+    }
+
+    @Override
+    public void run() {
+        try {
+            pollTxnStream();
+        } catch (Throwable throwable) {
+            StreamListener listener = subscription.getListener();
+            log.error("Encountered exception {} during txn stream polling, listener: {}, " +
+                    "namespace: {}", throwable, listener, subscription.getNamespace());
+            streamingManager.unsubscribe(listener, false);
+            listener.onError(throwable);
+        }
+    }
+
+    /**
+     * Poll new data changes from the transaction stream and put into
+     * the subscription's buffer one by one.
+     */
+    private void pollTxnStream() throws Exception {
+        // If listener already unsubscribed, do not process or schedule again.
+        if (subscription.isStopped()) {
+            return;
+        }
+
+        // Seek to next address and poll transaction updates.
+        txnStream.seek(lastReadAddress + 1L);
+        List<ILogData> updates = txnStream.remaining();
```

Comment:
use of remaining() is exactly what got us into the OOM scenario. Even with more selectivity, this will result in OOM if there is a lot of distance between the lastAddress and the tail.

Reply:
Initially I tried to add a parameter in remaining() to limit the number of entries returned, but the current implementation makes it complicated and error-prone to make this change. However, with the try catch block and the tunable buffer size, when OOM happens it will be captured and user be able to handle it by re-process the initial snapshot and re-subscribe. Let's sync up this and other concern you might have on Monday.

Reply:
OOM exceptions cannot always be caught in a try-catch block - the actual OOM error can happen in a stack trace not controlled by us.
Use of remaining() the way it is very buggy, if there are any infra issues then the stream address space can grow quickly which means remaining() will pull the entire stream history into memory.
Maybe this loop can be change to do a next() up to the NOTIFICATION_BATCH_SIZE and then stop?

Reply:
In initial impl, OOM can happen in a stack trace not controlled by us because it has large buffer size, but now we can fine tine it. Using next() is too slow and not a generic approach. And if you remember even if we use next(), they still have OOM because their jvm is too small, that problem is not one our side, we don't want to do our impl for a specific case that slows down other common use cases right? The right solution is too enhance remaining()/remainingUpto() API so that it limits the number of entries returned, the current maxGlobal parameter in streaming API is not the number of entries returned for this stream, but the last GLOBAL address to look for, which might return 0 stream entries up to that address, so need another parameter to control the number of actual entries returned.